File: python/ch19/mcts_toy.py
# python/ch19/mcts_toy.py
"""4 ノードグラフ上のターン制ゲームに MCTS（UCB1）を適用（第19章）."""
from __future__ import annotations

import math
import random
import sys
from dataclasses import dataclass, field
from typing import Dict, List, Optional, TextIO, Tuple

EXPLORATION = 1.41
ITERATIONS = 300
MAX_TURNS = 16


@dataclass
class GameInput:
    nodes: int
    edges: List[Tuple[int, int]]
    start: int
    terminal: int = 3


@dataclass
class GameState:
    node: int
    player: int


@dataclass
class MCTSNode:
    move: Optional[int] = None
    state: Optional[GameState] = None
    parent: Optional[MCTSNode] = None
    children: Dict[int, MCTSNode] = field(default_factory=dict)
    untried: List[int] = field(default_factory=list)
    visits: int = 0
    wins: float = 0.0


def read_game(stream: TextIO) -> GameInput:
    lines = stream.read().splitlines()
    parts = list(map(int, lines[0].split()))
    nodes, m = parts[0], parts[1]
    edges = []
    for i in range(1, 1 + m):
        u, v = map(int, lines[i].split())
        edges.append((u, v))
    start = int(lines[1 + m])
    return GameInput(nodes=nodes, edges=edges, start=start)


def build_adjacency(game: GameInput) -> List[List[int]]:
    adj: List[List[int]] = [[] for _ in range(game.nodes)]
    for u, v in game.edges:
        adj[u].append(v)
        adj[v].append(u)
    for i in range(game.nodes):
        adj[i].sort()
    return adj


def terminal_winner(state: GameState, game: GameInput) -> Optional[int]:
    if state.node == game.terminal:
        return state.player
    return None


def apply_move(state: GameState, nxt_node: int) -> GameState:
    return GameState(node=nxt_node, player=1 - state.player)


def rollout(state: GameState, adj: List[List[int]], game: GameInput, rng: random.Random) -> int:
    cur = state
    for _ in range(game.nodes * 4):
        winner = terminal_winner(cur, game)
        if winner is not None:
            return winner
        moves = adj[cur.node]
        if not moves:
            return 1 - cur.player
        cur = apply_move(cur, rng.choice(moves))
    return 1 - cur.player


def ucb1(parent_visits: int, child: MCTSNode) -> float:
    if child.visits == 0:
        return float("inf")
    return child.wins / child.visits + EXPLORATION * math.sqrt(math.log(parent_visits) / child.visits)


def expand(node: MCTSNode, adj: List[List[int]], rng: random.Random) -> MCTSNode:
    assert node.state is not None
    move = node.untried.pop(rng.randrange(len(node.untried)))
    child_state = apply_move(node.state, move)
    child = MCTSNode(move=move, state=child_state, parent=node, untried=list(adj[child_state.node]))
    node.children[move] = child
    return child


def best_child(node: MCTSNode) -> MCTSNode:
    return max(node.children.values(), key=lambda c: ucb1(node.visits, c))


def mcts(root: MCTSNode, adj: List[List[int]], game: GameInput, rng: random.Random) -> None:
    for _ in range(ITERATIONS):
        node = root
        path: List[MCTSNode] = [node]

        while not node.untried and node.children and node.state is not None:
            winner = terminal_winner(node.state, game)
            if winner is not None:
                break
            node = best_child(node)
            path.append(node)

        winner = None
        if node.state is not None:
            winner = terminal_winner(node.state, game)

        if winner is None and node.untried and node.state is not None:
            node = expand(node, adj, rng)
            path.append(node)

        if node.state is not None and winner is None:
            winner = rollout(node.state, adj, game, rng)
        elif winner is None:
            winner = 0

        for n in path:
            n.visits += 1
            if n.state is not None and winner == n.state.player:
                n.wins += 1.0


def select_move(root: MCTSNode, adj: List[List[int]]) -> int:
    if not root.children:
        return adj[root.state.node][0]  # type: ignore[union-attr]
    return max(root.children, key=lambda m: root.children[m].visits)


def play_game(game: GameInput, rng: random.Random) -> List[int]:
    adj = build_adjacency(game)
    state = GameState(node=game.start, player=0)
    log: List[int] = []

    for _ in range(MAX_TURNS):
        winner = terminal_winner(state, game)
        if winner is not None:
            break
        moves = adj[state.node]
        if not moves:
            break

        root = MCTSNode(state=state, untried=list(moves))
        mcts(root, adj, game, rng)
        mv = select_move(root, adj)
        log.append(mv)
        state = apply_move(state, mv)

    return log


def main() -> None:
    rng = random.Random(42)
    game = read_game(sys.stdin)
    moves = play_game(game, rng)
    print(f"mcts moves={len(moves)} path={' '.join(map(str, moves))}", file=sys.stderr)
    for m in moves:
        print(m)


if __name__ == "__main__":
    main()

File: go/ch19/mcts_toy.go
// go/ch19/mcts_toy.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"math/rand"
	"os"
	"strconv"
	"strings"
)

const (
	exploration = 1.41
	iterations  = 300
	maxTurns    = 16
)

type GameInput struct {
	Nodes    int
	Edges    [][2]int
	Start    int
	Terminal int
}

type GameState struct {
	Node   int
	Player int
}

type MCTSNode struct {
	Move     int
	HasMove  bool
	State    GameState
	HasState bool
	Parent   *MCTSNode
	Children map[int]*MCTSNode
	Untried  []int
	Visits   int
	Wins     float64
}

func readGame(r *bufio.Reader) (GameInput, error) {
	line, err := r.ReadString('\n')
	if err != nil {
		return GameInput{}, err
	}
	parts := parseInts(strings.TrimSpace(line))
	nodes, m := parts[0], parts[1]
	edges := make([][2]int, m)
	for i := 0; i < m; i++ {
		line, err = r.ReadString('\n')
		if err != nil {
			return GameInput{}, err
		}
		p := parseInts(strings.TrimSpace(line))
		edges[i] = [2]int{p[0], p[1]}
	}
	line, err = r.ReadString('\n')
	if err != nil {
		return GameInput{}, err
	}
	start, _ := strconv.Atoi(strings.TrimSpace(line))
	return GameInput{nodes, edges, start, 3}, nil
}

func parseInts(line string) []int {
	parts := strings.Fields(line)
	out := make([]int, len(parts))
	for i, p := range parts {
		out[i], _ = strconv.Atoi(p)
	}
	return out
}

func buildAdjacency(game GameInput) [][]int {
	adj := make([][]int, game.Nodes)
	for _, e := range game.Edges {
		adj[e[0]] = append(adj[e[0]], e[1])
		adj[e[1]] = append(adj[e[1]], e[0])
	}
	return adj
}

func terminalWinner(state GameState, game GameInput) (int, bool) {
	if state.Node == game.Terminal {
		return state.Player, true
	}
	return 0, false
}

func applyMove(state GameState, nxt int) GameState {
	return GameState{nxt, 1 - state.Player}
}

func rollout(state GameState, adj [][]int, game GameInput, rng *rand.Rand) int {
	cur := state
	for i := 0; i < game.Nodes*4; i++ {
		if w, ok := terminalWinner(cur, game); ok {
			return w
		}
		moves := adj[cur.Node]
		if len(moves) == 0 {
			return 1 - cur.Player
		}
		cur = applyMove(cur, moves[rng.Intn(len(moves))])
	}
	return 1 - cur.Player
}

func ucb1(parentVisits int, child *MCTSNode) float64 {
	if child.Visits == 0 {
		return math.Inf(1)
	}
	return child.Wins/float64(child.Visits) + exploration*math.Sqrt(math.Log(float64(parentVisits))/float64(child.Visits))
}

func expand(node *MCTSNode, adj [][]int, rng *rand.Rand) *MCTSNode {
	idx := rng.Intn(len(node.Untried))
	move := node.Untried[idx]
	node.Untried = append(node.Untried[:idx], node.Untried[idx+1:]...)
	childState := applyMove(node.State, move)
	child := &MCTSNode{
		Move: move, HasMove: true,
		State: childState, HasState: true,
		Parent: node,
		Children: map[int]*MCTSNode{},
		Untried:  append([]int{}, adj[childState.Node]...),
	}
	node.Children[move] = child
	return child
}

func bestChild(node *MCTSNode) *MCTSNode {
	var best *MCTSNode
	bestScore := -1.0
	for _, c := range node.Children {
		s := ucb1(node.Visits, c)
		if best == nil || s > bestScore {
			best = c
			bestScore = s
		}
	}
	return best
}

func mcts(root *MCTSNode, adj [][]int, game GameInput, rng *rand.Rand) {
	for it := 0; it < iterations; it++ {
		node := root
		path := []*MCTSNode{node}

		for len(node.Untried) == 0 && len(node.Children) > 0 && node.HasState {
			if _, ok := terminalWinner(node.State, game); ok {
				break
			}
			node = bestChild(node)
			path = append(path, node)
		}

		winner := -1
		if node.HasState {
			if w, ok := terminalWinner(node.State, game); ok {
				winner = w
			}
		}

		if winner < 0 && len(node.Untried) > 0 && node.HasState {
			node = expand(node, adj, rng)
			path = append(path, node)
		}

		if winner < 0 && node.HasState {
			winner = rollout(node.State, adj, game, rng)
		} else if winner < 0 {
			winner = 0
		}

		for _, n := range path {
			n.Visits++
			if n.HasState && winner == n.State.Player {
				n.Wins++
			}
		}
	}
}

func selectMove(root *MCTSNode, adj [][]int) int {
	if len(root.Children) == 0 {
		return adj[root.State.Node][0]
	}
	bestMove := -1
	bestVisits := -1
	for m, c := range root.Children {
		if c.Visits > bestVisits {
			bestVisits = c.Visits
			bestMove = m
		}
	}
	return bestMove
}

func playGame(game GameInput, rng *rand.Rand) []int {
	adj := buildAdjacency(game)
	state := GameState{game.Start, 0}
	var log []int

	for turn := 0; turn < maxTurns; turn++ {
		if _, ok := terminalWinner(state, game); ok {
			break
		}
		moves := adj[state.Node]
		if len(moves) == 0 {
			break
		}
		root := &MCTSNode{
			State: state, HasState: true,
			Children: map[int]*MCTSNode{},
			Untried:  append([]int{}, moves...),
		}
		mcts(root, adj, game, rng)
		mv := selectMove(root, adj)
		log = append(log, mv)
		state = applyMove(state, mv)
	}
	return log
}

func main() {
	rng := rand.New(rand.NewSource(42))
	game, err := readGame(bufio.NewReader(os.Stdin))
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	moves := playGame(game, rng)
	fmt.Fprintf(os.Stderr, "mcts moves=%d path=", len(moves))
	for i, m := range moves {
		if i > 0 {
			fmt.Fprint(os.Stderr, " ")
		}
		fmt.Fprint(os.Stderr, m)
	}
	fmt.Fprintln(os.Stderr)
	for _, m := range moves {
		fmt.Println(m)
	}
}
